### ISSUE TITLE
CPU Memory Leakage on TPU

### ISSUE THREAD

**Issue by zeyuwang615** [pytorch/xla#4015]
## Description
I find strange memory behaviour with updating layer wise learning rate decay on tpu. If the layer decay learning rate is set (`--layer-decay`), the memory usage just keep going up. For small models this is fine. For large models this may eventually lead to out-of-memory error and program crashes. However, if the layer decay learning rate is turned off, everything seems good. 

In short, if lr scheduler uses
```
for param_group, value in zip(self.optimizer.param_groups, values):
    param_group[self.param_group_field] = lr * param_group["lr_scale"]
```
instead of 
```
for param_group, value in zip(self.optimizer.param_groups, values):
    param_group[self.param_group_field] = lr
```
, the problem comes.

Another interesting thing is that even after the program is finished, the cpu memory usage is still over 100GB. I have to use `kill -9` to release the memory.

I have also find that turning on the layer decay learning rate will essentially triple the time taken by the first few steps on tpu pods. I think this has something to do with graph compilation. But I fail to understand why updating optimizer learning rate has anything to do with the computational graph.

**Expected behavior**
This strange behaviour of `--layer-decay` not only greatly slows down the training speed (on pods the first few steps take almost as much time as the rest and 99.9% steps), but also leads to program crashes after a long time training. Now I have to manually kill the program twice a day. It would be nice if one can use tpu smoothly, just like gpu training.

**Screenshots**
![cpu_memory](https://user-images.githubusercontent.com/26880977/190921782-c2b7d04e-ccc8-42de-b426-9f8176748438.JPG)

The first half is the result without `--layer-decay`, and the second half is the result with `--layer-decay`. It is clear that without `--layer-decay`, the memory usage is stable. On the other hand, when `--layer-decay` is turned on, the memory usage keeps going up until the memory limit is reached and program crashes.

**Environment (please complete the following information):**
This issue is found on multiple tpu vm setup 
 - tpu v3-8, with tpu software version tpu-vm-pt-1.12
 - tpu v3 pods, with tpu software version tpu-vm-pt-1.12
 - tpu v4 pods, with tpu software version v2-alpha-tpuv4

**Comment by JackCaoG** [comment]
One thing I want to point out is that `optimizer` update is also part of the graph. For pytorch/xla you can think everything(forward, backward, optimizer) is captured within graphs(ideally a single graph).

It is hard to debug a huge model obviouslly, I think what we should do is start with a small model and compare the graph generated by both case.

Can you export `XLA_SAVE_TENSORS_FMT='hlo' XLA_SAVE_TENSORS_FILE='/tmp/save1.hlo'` and rerun the small model and check out the HLO under `/tmp/save1.hlo`? You should see some difference in the HLO.

**Comment by zeyuwang615** [comment]
Yes, I do see some differences. But I have never used this feature before and don't know how to read this file. Could you please help me on that?

Here are the HLO files of a small model for a single run.

with layer decay.
[save1_super_tiny_layer_decay.hlo.txt](https://github.com/pytorch/xla/files/9614443/save1_super_tiny_layer_decay.hlo.txt)

without layer decay
[save2_no_layer_decay.hlo.txt](https://github.com/pytorch/xla/files/9614448/save2_no_layer_decay.hlo.txt)

**Comment by JackCaoG** [comment]
I will try to set some time next week to repo this issue.

**Comment by zeyuwang615** [comment]
Hi, It has been 11days. Have you got any chance to look at these files?

**Comment by zeyuwang615** [comment]
Here is a link to the issue I've posted on pytorch-image-models about this memory problem in xla, which might be useful:
https://github.com/rwightman/pytorch-image-models/issues/1470#issue-1376631179

**Comment by JackCaoG** [comment]
I think Ross is right, if you need to convert a tensor(which is lazy) value back to cpu as a scalar every step, this will incur a lot of overhead. 

If you print
```
import torch_xla.debug.metrics as met

print(met.metrics_report())
```

for both cases, I expect you see a lot more `_local_scalar_dense` and `transfer_from_server` with `layer decay`. I am not sure why memory usage will increase so much through, I might have a better idea if you can share the metrics for both case.

**Comment by zeyuwang615** [comment]
Hi, the metrics reports are listed here. These two files are generated by vit-large after running for 10 steps.

[metrics_layer_decay_large.txt](https://github.com/pytorch/xla/files/9762752/metrics_layer_decay_large.txt)
[metrics_no_layer_decay_large.txt](https://github.com/pytorch/xla/files/9762753/metrics_no_layer_decay_large.txt)

I don't see `_local_scalar_dense` and `transfer_from_server` with `layer decay`, but there are some ops with more count when `layer decay` is set
